libsql-server/src/http/user/listen.rs

```diff
@@ -0,0 +1,184 @@
+            .header("Location", primary_url + uri.path())
+            .body(BoxBody::default())
+            .unwrap());
+    }
+
+    // TODO: validate table
+    let table = mem::take(&mut query.table);
+    let actions = mem::take(&mut query.action);
+
+    let stream =
+        SubscriptionAggregator::new(state.namespaces.clone(), namespace, table, actions).await?;
+    Ok(JsonLines::new(stream).into_response())
+}
+
+static LAGGED_MSG: &str = "some changes were lost";
+
+type AggregatorState = HashMap<Action, u64>;
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "snake_case")]
+enum AggregatorEvent {
+    Error(&'static str),
+    #[serde(untagged)]
+    Changes(AggregatorState),
+}
+
+type AggregatorResult = Result<AggregatorEvent, Error>;
+type AggregatorPoll = Poll<Option<AggregatorResult>>;
+
+struct SubscriptionAggregator {
+    actions: Vec<Action>,
+    subscription: UpdateSubscription,
+    state: AggregatorState,
+    store: NamespaceStore,
+    namespace: NamespaceName,
+    table: String,
+    errored: bool,
+}
+
+impl SubscriptionAggregator {
+    async fn new(
+        store: NamespaceStore,
+        namespace: NamespaceName,
+        table: String,
+        actions: Vec<Action>,
+    ) -> crate::Result<Self> {
+        let subscription = store
+            .subscribe(namespace.clone(), table.clone())
+            .await
+            .unwrap();
+        Ok(Self {
+            actions,
+            subscription,
+            state: HashMap::new(),
+            store,
+            namespace,
+            table,
+            errored: false,
+        })
+    }
+}
+
+impl Drop for SubscriptionAggregator {
+    fn drop(&mut self) {
+        let namespace = mem::take(&mut self.namespace);
+        let table = mem::take(&mut self.table);
+        let store = self.store.clone();
+        tokio::spawn(async move { _ = store.unsubscribe(namespace, table).await });
+    }
+}
+
+impl Stream for SubscriptionAggregator {
+    type Item = AggregatorResult;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(poll) = self.poll_inner(cx) {
+                return poll;
+            }
+        }
+    }
+}
+
+impl SubscriptionAggregator {
+    fn poll_inner(self: &mut Self, cx: &mut Context) -> Option<AggregatorPoll> {
+        match self.subscription.inner.poll_next_unpin(cx) {
+            Poll::Pending => Some(Poll::Pending),
+            Poll::Ready(value) => match value {
+                None => Some(Poll::Ready(None)),
+                Some(result) => match result {
+                    Ok(item) => match item {
+                        BroadcastMsg::Change { action, .. } => self.register(action),
+                        BroadcastMsg::Rollback => self.clear(),
+                        BroadcastMsg::Commit => self.flush(),
+                    },
+                    Err(_) => self.error(),
+                },
+            },
+        }
+    }
```

Comment:
Why do you have a ton of manual poll implementation? Can you use `async_stream` instead?

Reply:
Because I need to aggregate several items from upstream into one (or maybe none) to send downstream.
Took a quick look at `async_stream` and couldn't figure out how to use it for this purpose. Do you have an example?

If this isn't a blocker, I'd rather not change this for now. On the next iteration, I want to nuke this. 
The broadcaster channel is the biggest perf bottleneck, so I want to do the aggregation before sending data through it.
This means we won't need to aggregate at the stream level, since we do it before.